pkg/services/user/userimpl/store.go

```diff
@@ -88,8 +88,18 @@ func (ss *sqlStore) Insert(ctx context.Context, cmd *user.User) (int64, error) {
 }
 
 func (ss *sqlStore) Get(ctx context.Context, usr *user.User) (*user.User, error) {
+	ret := &user.User{}
 	err := ss.db.WithDbSession(ctx, func(sess *db.Session) error {
-		exists, err := sess.Where("email=? OR login=?", usr.Email, usr.Login).Get(usr)
+		where := "email=? OR login=?"
+		login := usr.Login
+		email := usr.Email
+		if ss.cfg.CaseInsensitiveLogin {
+			where = "LOWER(email)=LOWER(?) OR LOWER(login)=LOWER(?)"
+			login = strings.ToLower(login)
+			email = strings.ToLower(email)
```

Comment:
nit: login and email are being lowered twice - in the SQL query and by `strings.ToLower`.

Reply:
I think that the index is not used if case insensitive is on:

```
mysql> EXPLAIN SELECT * FROM user WHERE LOWER(login) = LOWER('admin') OR LOWER(email) = 'Admin';
+----+-------------+-------+------------+------+---------------+------+---------+------+------+----------+-------------+
| id | select_type | table | partitions | type | possible_keys | key  | key_len | ref  | rows | filtered | Extra       |
+----+-------------+-------+------------+------+---------------+------+---------+------+------+----------+-------------+
|  1 | SIMPLE      | user  | NULL       | ALL  | NULL          | NULL | NULL    | NULL | 4407 |   100.00 | Using where |
+----+-------------+-------+------------+------+---------------+------+---------+------+------+----------+-------------+
1 row in set, 1 warning (0.01 sec)
```

while the respective query for case sensitive search is:
```
mysql> EXPLAIN SELECT * FROM user WHERE login = 'admin' OR email = 'admin';
+----+-------------+-------+------------+-------------+--------------------------------------------------------------------------+-------------------------------+---------+------+------+----------+---------------------------------------------------------+
| id | select_type | table | partitions | type        | possible_keys                                                            | key                           | key_len | ref  | rows | filtered | Extra                                                   |
+----+-------------+-------+------------+-------------+--------------------------------------------------------------------------+-------------------------------+---------+------+------+----------+---------------------------------------------------------+
|  1 | SIMPLE      | user  | NULL       | index_merge | UQE_user_login,UQE_user_email,IDX_user_login_email,user_email,user_login | UQE_user_login,UQE_user_email | 762,762 | NULL |    2 |   100.00 | Using union(UQE_user_login,UQE_user_email); Using where |
+----+-------------+-------+------------+-------------+--------------------------------------------------------------------------+-------------------------------+---------+------+------+----------+---------------------------------------------------------+
1 row in set, 1 warning (0.01 sec)
```

a solution would be to [add an index for LOWER(email) and LOWER(login)](https://use-the-index-luke.com/sql/where-clause/functions/case-insensitive-search) but this would require extra storage.

Reply:
we could live with a "temp" index for case insensitive on, as this will eventually be removed as we will force users login and username to be lowered in the future

Reply:
il add this to an issue and work on it, but it is not necessary for this to be merged. really well-spotted @papagian

Reply:
https://github.com/grafana/grafana/issues/67498

Reply:
@IevaVasiljeva any preference for 
```go
			where = "LOWER(email)=LOWER(?) OR LOWER(login)=LOWER(?)"
```

```go
			login = strings.ToLower(login)
			email = strings.ToLower(email)
			where = "LOWER(email)=? OR LOWER(login)=?"

```

Reply:
> tested to run without lowering the parameters but it is still not using the idx, the idx needs to be added

do you mean the `IDX_user_login_email` is not used? I think that's because of the `OR` operator
also how do you check if the index is used? maybe the execution plan is different if there are too few rows in the table

Reply:
```sql
EXPLAIN SELECT * FROM user WHERE LOWER(login) = 'admin' OR LOWER(email) = 'admin';

```
tested this query which is not using the idex

Reply:
yea it's still not using the idx 😢 
```
EXPLAIN SELECT * FROM user WHERE LOWER(login) = 'admin' OR LOWER(email) = 'admin';
```

| id | select_type | table | partitions | type | possible_keys | key | key_len | ref | rows | filtered | Extra       |
|----|-------------|-------|------------|------|---------------|-----|---------|-----|------|----------|-------------|
|  1 | SIMPLE      | user  |            | ALL  |               |     |         |     | 1001 |   100.00 | Using where |

Reply:
niceeee formatting 🗡️

Reply:
I think you need two additional indexes for this:
```
mysql> CREATE INDEX IDX_user_lower_login ON user ( (LOWER(login)) );
Query OK, 0 rows affected (0.08 sec)
Records: 0  Duplicates: 0  Warnings: 0

mysql> CREATE INDEX IDX_user_lower_email ON user ( (LOWER(email)) );
Query OK, 6301 rows affected (0.18 sec)
Records: 6301  Duplicates: 0  Warnings: 0

mysql> EXPLAIN SELECT * FROM user WHERE LOWER(login) = 'admin' OR LOWER(email) = 'admin';
+----+-------------+-------+------------+-------------+-------------------------------------------+-------------------------------------------+---------+------+------+----------+--------------------------------------------------------------------------+
| id | select_type | table | partitions | type        | possible_keys                             | key                                       | key_len | ref  | rows | filtered | Extra                                                                    |
+----+-------------+-------+------------+-------------+-------------------------------------------+-------------------------------------------+---------+------+------+----------+--------------------------------------------------------------------------+
|  1 | SIMPLE      | user  | NULL       | index_merge | IDX_user_lower_login,IDX_user_lower_email | IDX_user_lower_login,IDX_user_lower_email | 763,763 | NULL |    2 |   100.00 | Using sort_union(IDX_user_lower_login,IDX_user_lower_email); Using where |
+----+-------------+-------+------------+-------------+-------------------------------------------+-------------------------------------------+---------+------+------+----------+--------------------------------------------------------------------------+
1 row in set, 1 warning (0.01 sec)
```